packages/angular/ssr/node/src/app-engine.ts

```diff
@@ -0,0 +1,95 @@
+import type { IncomingMessage } from 'http';
+import { createWebRequestFromNodeRequest } from './request';
+
+/**
+ * Angular server application engine.
+ * Manages Angular server applications (including localized ones) and handles rendering requests.
+
+ * @developerPreview
+ */
+export interface AngularNodeServerAppManager {
+  /**
+   * Renders an HTTP response based on the incoming request using the Angular server application.
+   *
+   * The method processes the incoming request, determines the appropriate route, and prepares the
+   * rendering context to generate a response. If the request URL corresponds to a static file (excluding `/index.html`),
+   * the method returns `null`.
+   *
+   * Example: A request to `https://www.example.com/page/index.html` will render the Angular route
+   * associated with `https://www.example.com/page`.
+   *
+   * @param request - The incoming HTTP request object to be rendered. It can be a `Request` or `IncomingMessage` object.
+   * @param requestContext - Optional additional context for the request, such as metadata or custom settings.
+   * @returns A promise that resolves to a `Response` object, or `null` if the request URL is for a static file
+   * (e.g., `./logo.png`) rather than an application route.
+   */
+  render(request: Request | IncomingMessage, requestContext?: unknown): Promise<Response | null>;
+}
+
+/**
+ * Angular server application engine.
+ * Manages Angular server applications (including localized ones), handles rendering requests,
+ * and optionally transforms index HTML before rendering.
+ */
+export class AngularNodeAppEngine extends AngularAppEngine implements AngularNodeServerAppManager {
+  /**
+   * Renders an HTTP response based on the incoming request using the Angular server application.
+   *
+   * The method processes the incoming request, determines the appropriate route, and prepares the
+   * rendering context to generate a response. If the request URL corresponds to a static file (excluding `/index.html`),
+   * the method returns `null`.
+   *
+   * Example: A request to `https://www.example.com/page/index.html` will render the Angular route
+   * associated with `https://www.example.com/page`.
+   *
+   * @param request - The incoming HTTP request object to be rendered. It can be a `Request` or `IncomingMessage` object.
+   * @param requestContext - Optional additional context for the request, such as metadata or custom settings.
+   * @returns A promise that resolves to a `Response` object, or `null` if the request URL is for a static file
+   * (e.g., `./logo.png`) rather than an application route.
+   */
+  override render(
+    request: Request | IncomingMessage,
+    requestContext?: unknown,
+  ): Promise<Response | null> {
+    const webReq = request instanceof Request ? request : createWebRequestFromNodeRequest(request);
+
+    return super.render(webReq, requestContext);
+  }
+}
+
+let angularNodeAppEngine: AngularNodeAppEngine | undefined;
```

Comment:
Question: I'm a little worried by the amount of singletons we've been introducing. I get that this is somewhat necessary to support injecting the manifest data and maybe it's unavoidable, but I'm wondering if there's any strategies we can use to at least minimize the usage of singletons. I assume this needs to be a singleton because `AngularAppEngine` is.

Does the CLI need knowledge of this many different symbols to inject the correct data? Is there a way we can consolidate onto a different abstraction which the CLI interacts with that doesn't require all the underlying pieces to be singletons as well?

Reply:
`AngularAppEngine` and `AngularNodeAppEngine` don't strictly need to be singletons, but creating multiple instances wouldn't be ideal because there is only one manifest corresponding to a single app.

The `getOrCreateX` and `destroyX` methods were implemented to clearly convey this to users. 

The CLI will only interact directly with the `AngularAppEngine` class.

// @AndrewKushnir, these methods were created based on his feedback.

Reply:
Is there a significant risk of developers using multiple `AngularAppEngine` classes? For the most part, users are going to copy whatever we scaffold, and any cloud integrations would likely take the effort to reuse something like this. We could potentially also do something like `new AngularAppEngine({ manifest: globalThis.ngSsrManifest })` to manually link this to the manifest singleton rather than forcing it to be a singleton too (maybe it just defaults to that).

Reply:
> Is there a significant risk of developers using multiple AngularAppEngine classes?

I don't think the risk is high, but when it happens, what would be the behavior? There should also be a way to retrieve a reference to a singleton instance itself, in case it was created in a different part of the code (potentially not owned by an app developer). We can potentially come up with different APIs to represent that if the existing ones don't look ergonomic.

Reply:
> I don't think the risk is high, but when it happens, what would be the behavior? 

Potentially the impact would be some very minor performance impacts.

> In case it was created in a different part of the code (potentially not owned by an app developer).

I'm not entirely clear on the use-case you're describing.

I'm not particularly attached to any specific approach, but we should definitely decide on the desired structure of the public API.